tests/regress/hook_code_stop_emu.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import print_function
+import binascii
+import regress
+
+from unicorn import *
+from unicorn.x86_const import *
+
+
+CODE = binascii.unhexlify(b"".join([
+  b"48c7c003000000", # mov rax, 3      mapped: 0x1000
+  b"0f05",           # syscall         mapped: 0x1007
+  b"48c7c700400000", # mov rdi, 0x4000 mapped: 0x1009
+  b"488907",         # mov [rdi], rdx  mapped: 0x1010
+  b"488b07",         # mov rdx, [rdi]  mapped: 0x1013
+  b"4883c201",       # add rdx, 1      mapped: 0x1016
+  ]))
+
+
+class SingleStepper:
+    def __init__(self, emu, test):
+        self._emu = emu
+        self._hit_count = 0
+        self._test = test
+
+    def _stop_hook(self, uc, address, *args, **kwargs):
+        if self._hit_count == 0:
+            self._hit_count += 1
+        else:
+            self._test.assertEqual(1, self._hit_count, "HOOK_CODE invoked too many times")
+            uc.emu_stop()
```

Comment:
Sorry, I should have documented the code.

The intent of the `.step()` method is to emulate exactly one instruction. Because the `HOOK_CODE` callback is invoked before an instruction is emulated, we need to ignore the first invocation (line 27), and then stop the emulator before any subsequent instructions (line 30).

Looks like I missed a `self._hit_count += 1` after line 30, but that doesn't change the functionality of the method. Just the assertion on line 29 (`_hit_count` will never be greater than 1 right now).